src/registry/maven.rs

```diff
@@ -0,0 +1,446 @@
+                    .into_diagnostic()
+                    .wrap_err(miette!("failed to parse existing maven-metadata.xml"))?
+            } else {
+                // Create new metadata
+                MavenMetadata::new(package_name.to_string())
+            }
+        } else {
+            // Create new metadata
+            MavenMetadata::new(package_name.to_string())
+        };
+
+        // Add the new version if it doesn't exist
+        let version_str = new_version.to_string();
+        if !metadata.versioning.versions.contains(&version_str) {
+            metadata.versioning.versions.push(version_str.clone());
+            // Sort versions
+            metadata.versioning.versions.sort_by(|a, b| {
+                let va = Version::parse(a).ok();
+                let vb = Version::parse(b).ok();
+                match (va, vb) {
+                    (Some(va), Some(vb)) => va.cmp(&vb),
+                    _ => a.cmp(b),
+                }
+            });
+        }
+
+        // Update latest version
+        metadata.versioning.latest = Some(version_str.clone());
+        metadata.versioning.release = Some(version_str);
+
+        // Update lastUpdated timestamp
+        let now = chrono::Utc::now();
+        metadata.versioning.last_updated = Some(now.format("%Y%m%d%H%M%S").to_string());
+
+        // Serialize to XML
+        let xml = quick_xml::se::to_string(&metadata)
+            .into_diagnostic()
+            .wrap_err(miette!("failed to serialize maven-metadata.xml"))?;
+
+        // Upload the updated metadata
+        self.new_request(Method::PUT, metadata_url)
+            .body(xml)
+            .send()
+            .await?;
+
+        tracing::debug!("Updated maven-metadata.xml for {}", package_name);
+
+        Ok(())
+    }
+}
+
+struct RequestBuilder(reqwest::RequestBuilder);
+
+impl RequestBuilder {
+    fn new(client: reqwest::Client, method: reqwest::Method, url: Url) -> Self {
+        Self(client.request(method, url))
+    }
+
+    fn auth(mut self, token: String) -> Self {
+        self.0 = self.0.bearer_auth(token);
+        self
+    }
+
+    fn body(mut self, payload: impl Into<Body>) -> Self {
+        self.0 = self.0.body(payload);
+        self
+    }
+
+    async fn send(self) -> miette::Result<ValidatedResponse> {
+        self.0.send().await.into_diagnostic()?.try_into()
+    }
+}
+
+#[derive(Debug)]
+struct ValidatedResponse(reqwest::Response);
+
+impl TryFrom<Response> for ValidatedResponse {
+    type Error = miette::Report;
+
+    fn try_from(value: Response) -> Result<Self, Self::Error> {
+        ensure!(
+            !value.status().is_redirection(),
+            "remote server attempted to redirect request - is this registry URL valid?"
+        );
+
+        ensure!(
+            value.status() != 401,
+            "unauthorized - please provide registry credentials with `buffrs login`"
+        );
+
+        value.error_for_status().into_diagnostic().map(Self)
+    }
+}
```

Comment:
Might make sense to share `RequestBuilder` and `ValidatedResponse` between artifactory and maven?